aten/src/ATen/TensorIterator.cpp

```diff
@@ -830,12 +830,12 @@ bool TensorIteratorBase::is_cpu_scalar(int arg) const {
 void TensorIteratorBase::cast_outputs() {
   for (auto& op : operands_) {
     if (op.is_output && op.original_tensor_base().defined() &&
-        op.original_tensor_base().scalar_type() != op.current_dtype) {
+        !op.original_tensor_base().is_same(op.tensor_base())) {
```

Comment:
This doesn't look semantics preserving

Reply:
Hi, Yang. There is no any effect to any devices. But using `is_same` function to check tensors stored in operand, it may be more universal. original_tensor_base and tensor_base are not only inconsistent in data type but also possibly in strides and so on. Checking through tensor_impl allows for use by third-party devices. As for CPU, GPU, or other devices, such modifications have no impact.
1) For GPU devices and other devices, this function is not using.
2) For CPU devices, `!op.original_tensor_base().is_same(op.tensor_base())` is same with `p.original_tensor_base().scalar_type() != op.current_dtype`(op.current_dtype is same as tensor_base scalar_type) . Because output operand maybe call exchange tensor(original_tensor_base store old scalar type tensor, and tensor base store new scalar type tensor.) and set operand current dtype to common dtype when operand current dtype is not same with common_dtype_. https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/TensorIterator.cpp#L520

Reply:
But we don't need to cast the outputs if they're inconsistent in strides (and also they shouldn't be inconsistent in strides anyway, we should have set that up properly). The point of this function is to get the dtype correct...

Reply:
Ok, This can be done in custom device side. It has been restored.